File: data-runtime/internal/apps/aims/product_handoff_runtime.go
package aims

import (
	"context"
	"database/sql"
	"net/http"
	"net/url"

	"github.com/huizhi-yun/data-runtime/internal/apps/aims/productcenter"
	"github.com/huizhi-yun/data-runtime/internal/httperror"
)

func (a *Adapter) handleProductHandoffRuntime(ctx context.Context, method, path string, query url.Values, body map[string]any) (any, string, bool, error) {
	for _, suffix := range []string{"handoff-project:authorization", "planning-handoff:create"} {
		code, match := pathParam(path, "/v1/aims/internal/products/", "/"+suffix)
		if !match {
			continue
		}
		operation := "aims.product-priorities." + suffix
		if method != http.MethodPost {
			return nil, operation, true, httperror.New(405, "method_not_allowed", "POST required")
		}
		capability := "aims:product-priorities:handoff"
		if suffix == "handoff-project:authorization" {
			capability = "aims:product-priorities:project-authorization"
		}
		if err := requireProductServiceCapability(query, capability); err != nil {
			return nil, operation, true, err
		}
		uid := query.Get("current_user")
		if suffix == "handoff-project:authorization" {
			var input struct {
				ProjectCode string `json:"project_code"`
			}
			if err := decodeProductCommandPart(body["input"], &input); err != nil {
				return nil, operation, true, err
			}
			facts, err := loadProductHandoffProjectFacts(ctx, a.DB(), input.ProjectCode, uid)
			return facts, operation, true, productRuntimeError(err)
		}
		var input productcenter.PlanningHandoffInput
		var planningPermit, requestPermit, versionPermit productcenter.AuthorizationPermit
		var projectPermit productHandoffProjectPermit
		for _, part := range []struct {
			key    string
			target any
		}{{"input", &input}, {"planning_authorization", &planningPermit}, {"project_authorization", &projectPermit}} {
			if err := decodeProductCommandPart(body[part.key], part.target); err != nil {
				return nil, operation, true, err
			}
		}
		if input.RequestBizID != "" {
			if err := decodeProductCommandPart(body["request_authorization"], &requestPermit); err != nil {
				return nil, operation, true, err
			}
		}
		if input.PlannedVersionID > 0 {
			if err := decodeProductCommandPart(body["version_authorization"], &versionPermit); err != nil {
				return nil, operation, true, err
			}
		}
		var projectID int64
		target := productcenter.PlanningHandoffTarget{
			AuthorizeProject: func(ctx context.Context, tx *sql.Tx) (int64, error) {
				if input.PlannedVersionID > 0 {
					if err := productcenter.AuthorizeWorkspaceTransaction(ctx, tx, code, uid, "product_versions", "view", versionPermit); err != nil {
						return 0, err
					}
				}
				var err error
				projectID, err = authorizeProductHandoffProjectTx(ctx, tx, input.ProjectCode, uid, projectPermit)
				return projectID, err
			},
			ResolveRequirement: func(ctx context.Context, tx *sql.Tx) (int64, error) {
				return a.resolveProductHandoffRequirementTx(ctx, tx, code, uid, projectID, input)
			},
		}
		key, _ := body["idempotency_key"].(string)
		result, err := productcenter.HandoffPlanningItem(ctx, a.DB(), productcenter.CommandIdentity{ProductCode: code, ActorUID: uid, Action: "product_priorities:handoff", IdempotencyKey: key}, planningPermit, requestPermit, input, target)
		return result, operation, true, productRuntimeError(err)
	}
	return nil, "", false, nil
}

File: aims/server/utils/productHandoffRuntime.ts
import { createError, getHeader, getQuery, getRouterParam, readBody, setHeader, type H3Event } from 'h3'
import { maybeCallTenantRuntime } from '@hzy/foundation/server/utils/tenantRuntimeClient'
import { requireProductPermission } from './productAuthorization'
import { checkAimsScopedPermission } from './aimsScopedAuthorization'
import { productHandoffProjectObject, type ProductHandoffProjectFacts } from './productHandoffProjectAuthorizationCore'
import { productHandoffInput } from './productHandoffInput'
import { hasProductControlCharacter, productCommandKey } from './productWorkspaceInput'
import { runtimeEnvelopeError } from './aimsRuntimeForward'

export async function handleProductHandoff(event: H3Event, source: 'planning' | 'request') {
  setHeader(event, 'Cache-Control', 'no-store')
  const code = getRouterParam(event, 'productCode') || ''
  const key = productCommandKey(getHeader(event, 'Idempotency-Key'))
  if (!key || !code || code !== code.trim() || [...code].length > 64 || code.includes('/') || hasProductControlCharacter(code) || Object.keys(getQuery(event)).length) throw createError({ statusCode: 400, message: '转交路径或幂等键无效' })
  const raw = await readBody<unknown>(event)
  let body = raw
  let itemId = getRouterParam(event, 'itemId') || ''
  if (source === 'request') {
    if (!raw || typeof raw !== 'object' || Array.isArray(raw)) throw createError({ statusCode: 400, message: '转交参数无效' })
    const { planningItemId, ...rest } = raw as Record<string, unknown>
    if (typeof planningItemId !== 'string') throw createError({ statusCode: 400, message: '请选择对应的规划事项' })
    itemId = planningItemId
    body = rest
  }
  const input = productHandoffInput(body, itemId, source === 'request' ? getRouterParam(event, 'requestId') || '' : undefined)
  if (!input) throw createError({ statusCode: 400, message: '转交参数无效' })
  const planningFacts = await requireProductPermission(event, code, 'product_priorities', 'handoff')
  const requestFacts = input.request_biz_id ? await requireProductPermission(event, code, 'product_requests', 'handoff') : null
  const versionFacts = input.planned_version_id ? await requireProductPermission(event, code, 'product_versions', 'view') : null
  const uid = planningFacts.actor_uid
  const project = await maybeCallTenantRuntime<{ code: number, data: ProductHandoffProjectFacts }>(event, `/v1/aims/internal/products/${encodeURIComponent(code)}/handoff-project:authorization`, {
    appCode: 'aims', method: 'POST', scope: 'aims.read aims:product-priorities:project-authorization', query: { current_user: uid }, body: { input: { project_code: input.project_code } }
  })
  if (!project.handled) throw createError({ statusCode: 503, message: '项目授权数据暂不可用' })
  if (project.data.code !== 0) throw runtimeEnvelopeError(project.data)
  const object = productHandoffProjectObject(project.data.data, input.project_code, uid)
  if (!object) throw createError({ statusCode: 503, message: '项目授权上下文不一致' })
  if (!await checkAimsScopedPermission(event, { resourceCode: 'requirements', action: 'edit', object })) throw createError({ statusCode: 403, message: '没有目标项目的需求编辑权限' })
  const expires = Date.now() + 15000
  const runtime = await maybeCallTenantRuntime<{ code: number, data: unknown }>(event, `/v1/aims/internal/products/${encodeURIComponent(code)}/planning-handoff:create`, {
    appCode: 'aims', method: 'POST', scope: 'aims.write aims:product-priorities:handoff', query: { current_user: uid }, idempotencyKey: key,
    body: {
      input,
      planning_authorization: { resource: 'product_priorities', action: 'handoff', facts: planningFacts, expires_at: expires },
      project_authorization: { resource: 'requirements', action: 'edit', facts: project.data.data, expires_at: expires },
      ...(requestFacts ? { request_authorization: { resource: 'product_requests', action: 'handoff', facts: requestFacts, expires_at: expires } } : {}),
      ...(versionFacts ? { version_authorization: { resource: 'product_versions', action: 'view', facts: versionFacts, expires_at: expires } } : {})
    }
  })
  if (!runtime.handled) throw createError({ statusCode: 503, message: '产品运行服务暂不可用' })
  if (runtime.data.code !== 0) throw runtimeEnvelopeError(runtime.data)
  return runtime.data
}
